packages/commerce-sdk-react/src/hooks/ShopperBaskets/query.ts

```diff
@@ -6,73 +6,219 @@
+    const {headers, rawResponse, ...parameters} = arg
+    const defaultOptions = {
+        enabled: !!parameters.basketId
+    }
+    return useQuery(
+        ['/baskets', parameters.basketId, '/price-books', arg],
+        (_, {shopperBaskets}) =>
+            shopperBaskets.getPriceBooksForBasket({parameters, headers}, rawResponse),
+        {...defaultOptions, ...options}
+    )
 }
+
+type UseShippingMethodsForShipmentParameters = NonNullable<
+    Argument<Client['getShippingMethodsForShipment']>
+>['parameters']
+type UseShippingMethodsForShipmentHeaders = NonNullable<
+    Argument<Client['getShippingMethodsForShipment']>
+>['headers']
+type UseShippingMethodsForShipmentArg = {
+    headers?: UseShippingMethodsForShipmentHeaders
+    rawResponse?: boolean
+} & UseShippingMethodsForShipmentParameters
 /**
  * A hook for `ShopperBaskets#getShippingMethodsForShipment`.
  * Gets the applicable shipping methods for a certain shipment of a basket.
  * @see {@link https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-baskets?meta=getShippingMethodsForShipment} for more information about the API endpoint.
  * @see {@link https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperbaskets.shopperbaskets-1.html#getshippingmethodsforshipment} for more information on the parameters and returned data type.
  * @returns An object describing the state of the request.
  */
-export const useShippingMethodsForShipment = (
-    arg: Argument<Client['getShippingMethodsForShipment']>
-): UseQueryResult<DataType<Client['getShippingMethodsForShipment']>, Error> => {
-    const {shopperBaskets: client} = useCommerceApi()
-    return useQuery(['shipping-methods'], () => client.getShippingMethodsForShipment(arg))
+function useShippingMethodsForShipment(
+    arg: Omit<UseShippingMethodsForShipmentArg, 'rawResponse'> & {rawResponse?: false},
+    options?: UseQueryOptions<DataType<Client['getShippingMethodsForShipment']> | Response, Error>
+): UseQueryResult<DataType<Client['getShippingMethodsForShipment']>, Error>
+function useShippingMethodsForShipment(
+    arg: Omit<UseShippingMethodsForShipmentArg, 'rawResponse'> & {rawResponse: true},
+    options?: UseQueryOptions<DataType<Client['getShippingMethodsForShipment']> | Response, Error>
+): UseQueryResult<Response, Error>
+function useShippingMethodsForShipment(
+    arg: UseShippingMethodsForShipmentArg,
+    options?: UseQueryOptions<DataType<Client['getShippingMethodsForShipment']> | Response, Error>
+): UseQueryResult<DataType<Client['getShippingMethodsForShipment']> | Response, Error> {
+    const {headers, rawResponse, ...parameters} = arg
+    const defaultOptions = {
+        enabled: !!parameters.basketId && !!parameters.shipmentId
+    }
+    return useQuery(
+        [
+            '/baskets',
+            parameters.basketId,
+            '/shipments',
+            parameters.shipmentId,
+            '/shipping-methods',
+            arg
+        ],
```

Comment:
Some thoughts around the query key.

1️⃣ **Use familiar App concepts as query key**
The proposed query key is taking the perspective from the SCAPI endpoints.

But potentially users would like to take a more practical perspective by using App concepts to have a more fluent development experience.

Examples:

1. List of product per page showed on each product list page (PDP). E.g. `[{entity:'product', pageNum:{ page }...]`

1. Scope all data related to a registered user under the same scope. That will simplify to one call removing registerd user data from cache in cases like when the users logs out. E.g. `[{scope:'registeredUser'}...]`

2️⃣ **Scope `commerce-sdk-react` query key to avoid collision with external APIs user space query keys**
Today we are sharing the same query cache between `commerce-sdk-react` and the user land.
The query keys will be visible by users, and users will need to interact with the same cache space. 
I think users can accidentally access the wrong query cache entries if we decide to use fixed array positions in the query key.
We've to be mindful with users that disagree with our opinionated query key and leave them room to implement the queries to externals apis with their own flavour.

3️⃣ **Use one object as query key**
Using only one object inside an array provides more flexibility to work with the cache with fuzzy matching. 
We don't need worry about the specific position of the elements in the  query key array. Complicated query keys like this one, users will need to escape multiple array positions to address the scope they need.
E.g. Invalidate a shipping-method.
 `['','','','','/shipping-methods', 'finally-the-shipping-method-id]`
It's easy to address the wrong cache entry if you miss one of those empty strings in the array.

More details about using a query key with only one object: https://tkdodo.eu/blog/leveraging-the-query-function-context#object-query-keys

4️⃣ **Scope concepts using a different criteria**
In this example, we're tying together a shipping method to a specific basket, whereas shipping method in Business Manager are probably scoped by region not per basket.

Reply:
Thanks for sharing the thoughts Adam! I'm going to share my thoughts on each of the items you listed.

1. Re. Use familiar App concepts as query key

First of all, the cache keys must be one of the parameters of the API endpoint otherwise it defeats the purpose of being a cache key.

If we could come up with a schema, or a definition of what are the app concepts, and the corresponding relationship between those concept to the API parameters, and it would be great to share that to developers get some feedback. 

But even thou, we are introducing "App Concepts" as an API, in your example, we call it "pageNum" or "entity" or "scope", they could mean a lot to you, but it could also be meaningless to another developer. You are trying to establish a brand new pattern for developers to learn and I'm just worried that's not going to encourage developers to use this lib. 

Alternatively, if you see a RESTful endpoint path `/product` you immediately understand that product is an entity.

2. Collision

We could namespace the keys. This is a valid concern but I think the possibility of running into this is very very low and I'd love to deprioritize this issue, if you agree with me.

3. Object as key

It's important to note that we are comparing two **APIs designs** here:

- use restful endpoint path as array position
- come up with an object that has specific keys

I actually think using path as array is a benefit not a disadvantage, the array gives you information about the endpoint and the relationship between entities.

What would the cache key looks like for this endpoint? `/baskets/{basketId}/shipments/{shipmentId}/shipping-methods`

4. scopes
🤔

Reply:
@adamraya what would be the cache key for the endpoint (/baskets/{basketId}/shipments/{shipmentId}/shipping-methods) look like in the object approach?